python/acquire-zarr-py.cpp

```diff
@@ -0,0 +1,443 @@
+            compression_settings.codec = settings.compression()->codec();
+            compression_settings.level = settings.compression()->level();
+            compression_settings.shuffle = settings.compression()->shuffle();
+            stream_settings.compression_settings = &compression_settings;
+        }
+
+        const auto& dims = settings.dimensions();
+
+        std::vector<ZarrDimensionProperties> dimension_props;
+        std::vector<std::string> dimension_names(dims.size());
+        for (auto i = 0; i < dims.size(); ++i) {
+            const auto& dim = dims[i];
+            dimension_names[i] = dim.name();
+            ZarrDimensionProperties properties{
+                .name = dimension_names[i].c_str(),
+                .type = dim.type(),
+                .array_size_px = dim.array_size_px(),
+                .chunk_size_px = dim.chunk_size_px(),
+                .shard_size_chunks = dim.shard_size_chunks(),
+            };
+            dimension_props.push_back(properties);
+        }
+
+        stream_settings.dimensions = dimension_props.data();
+        stream_settings.dimension_count = dims.size();
+
+        stream_ =
+          ZarrStreamPtr(ZarrStream_create(&stream_settings), ZarrStreamDeleter);
+        if (!stream_) {
+            PyErr_SetString(PyExc_RuntimeError, "Failed to create Zarr stream");
+            throw py::error_already_set();
+        }
+    }
+
+    ~PyZarrStream()
+    {
+        if (is_active()) {
+            ZarrStream_destroy(stream_.get());
+        }
+    }
+
+    void append(py::array image_data)
+    {
+        if (!is_active()) {
+            PyErr_SetString(PyExc_RuntimeError,
+                            "Cannot append unless streaming.");
+            throw py::error_already_set();
+        }
+
+        auto buf = image_data.request();
+        auto* ptr = (uint8_t*)buf.ptr;
+
+        size_t bytes_out;
+        auto status = ZarrStream_append(
+          stream_.get(), ptr, buf.itemsize * buf.size, &bytes_out);
+
+        if (status != ZarrStatusCode_Success) {
+            std::string err = "Failed to append data to Zarr stream: " +
+                              std::string(Zarr_get_status_message(status));
+            PyErr_SetString(PyExc_RuntimeError, err.c_str());
+            throw py::error_already_set();
+        }
+    }
+
+    bool is_active() const { return static_cast<bool>(stream_); }
+
+  private:
+    using ZarrStreamPtr =
+      std::unique_ptr<ZarrStream, decltype(ZarrStreamDeleter)>;
+
+    ZarrStreamPtr stream_;
+};
+
+PYBIND11_MODULE(acquire_zarr, m)
+{
+    m.doc() = R"pbdoc(
+        Acquire Zarr Writer Python API
+        -----------------------
+        .. currentmodule:: acquire_zarr
+        .. autosummary::
+           :toctree: _generate
+           append
+    )pbdoc";
+
+    py::enum_<ZarrVersion>(m, "Version")
+      .value("V2", ZarrVersion_2)
+      .value("V3", ZarrVersion_3)
+      .export_values();
+
+    py::enum_<ZarrDataType>(m, "DType")
```

Comment:
A topic for discussion: do we think it might be worth building ways to implicitly (or explicitly) convert between numpy's DType & ZarrDataTypes?

Reply:
Yeah, I think that might be worthwhile to do.